tests/Integration/WhereConditions/WhereHasConditionsDirectiveTest.php

```diff
@@ -0,0 +1,471 @@
+        {
+            companies(
+                hasUser: {
+                    column: "name"
+                    value: ""
+                }
+            ) {
+                id
+                name
+            }
+        }
+        ')->assertJson([
+            'data' => [
+                'companies' => [
+                    [
+                        'id' => $userNamedEmptyString->company_id,
+                    ],
+                ],
+            ],
+        ]);
+    }
+
+    public function testCanQueryForNull(): void
+    {
+        factory(User::class, 3)->create();
+
+        $userNamedNull = factory(User::class)->create([
+            'name' => null,
+        ]);
+
+        $this->graphQL(/** @lang GraphQL */ '
+        {
+            companies(
+                hasUser: {
+                    column: "name"
+                    value: null
+                }
+            ) {
+                id
+            }
+        }
+        ')->assertJson([
+            'data' => [
+                'companies' => [
+                    [
+                        'id' => $userNamedNull->company_id,
+                    ],
+                ],
+            ],
+        ]);
+    }
+
+    public function testRequiresAValueForAColumn(): void
+    {
+        $this->graphQL(/** @lang GraphQL */ '
+        {
+            users(
+                hasPost: {
+                    column: "no_value"
+                }
+            ) {
+                id
+            }
+        }
+        ')->assertJsonFragment([
+            'message' => SQLOperator::missingValueForColumn('no_value'),
+        ]);
+    }
+
+    public function testOnlyAllowsWhitelistedColumns(): void
+    {
+        factory(User::class)->create();
+
+        $this->graphQL(/* @lang GraphQL */ '
+        {
+            whitelistedColumns(
+                hasCompany: {
+                    column: ID
+                    value: 1
+                }
+            ) {
+                id
+            }
+        }
+        ')->assertJson([
+            'data' => [
+                'whitelistedColumns' => [
+                    [
+                        'id' => 1,
+                    ],
+                ],
+            ],
+        ]);
+
+        $expectedEnumName = 'WhitelistedColumnsHasCompanyColumn';
+        $enum = $this->introspectType($expectedEnumName);
+
+        $this->assertArraySubset(
+            [
+                'kind' => 'ENUM',
+                'name' => $expectedEnumName,
+                'description' => 'Allowed column names for the `hasCompany` argument on the query `whitelistedColumns`.',
+                'enumValues' => [
+                    [
+                        'name' => 'ID',
+                    ],
+                    [
+                        'name' => 'CAMEL_CASE',
+                    ],
+                ],
+            ],
+            $enum
+        );
+    }
+
+    public function testWithoutRelationName(): void
```

Comment:
I think we really only need this test and maybe one or two of the other tests in this class.

Let's make sure the implementation is DRY so we can test it once and know it works anywhere.

Reply:
I abstracted as much as possible, no more repetitions. I'll probably be able to remove most of the duplicate tests now.

Reply:
I removed the redundant tests and added 2 new tests specific to the `@whereHasConditions` behavior.